plugin/trino-hive/src/main/java/io/trino/plugin/hive/TableChangeCallback.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive;
+
+import io.trino.plugin.hive.metastore.Partition;
+import io.trino.plugin.hive.metastore.Table;
+
+public interface TableChangeCallback
+{
+    TableChangeCallback NOOP = new TableChangeCallback()
+    {
+        @Override
+        public void onUpdateTable(Table table, boolean completedSuccessfully)
+        {
+        }
+
+        @Override
+        public void onUpdatePartition(Partition partition, boolean completedSuccessfully)
+        {
+        }
+
+        @Override
+        public void onDropTable(Table table, boolean completedSuccessfully)
+        {
+        }
+
+        @Override
+        public void onDropPartition(Partition partition, boolean completedSuccessfully)
+        {
+        }
+
+        @Override
+        public void onAlterTable(Table table, boolean completedSuccessfully)
+        {
+        }
+
+        @Override
+        public void onAlterPartition(Partition partition, boolean completedSuccessfully)
+        {
+        }
+    };
+
+    /**
+     * This operation is to be called after attempting to apply DML operations
+     * on the specified table.
+     */
+    void onUpdateTable(Table table, boolean completedSuccessfully);
+
+    /**
+     * This operation is to be called after attempting to apply DML operations
+     * on the specified partition.
+     */
+    void onUpdatePartition(Partition partition, boolean completedSuccessfully);
+
+    /**
+     * This operation is to be called after dropping a table.
+     */
+    void onDropTable(Table table, boolean completedSuccessfully);
+
+    /**
+     * This operation is to be called after dropping a partition.
+     */
+    void onDropPartition(Partition partition, boolean completedSuccessfully);
+
+    /**
+     * This operation is to be called after making structural changes on the
+     * specified table.
+     */
+    void onAlterTable(Table table, boolean completedSuccessfully);
```

Comment:
tracking the success/committed booleans to provide the value for the completedSuccessfully parameter introduces a lot of noise in the code.
the boolean is actually unused, but we're doing this to prevent us from lying that table is altered.

note however

- the implementation doesn't know what kind of table alteration happened, so it cannot draw much conclusions only from the fact `onAlterTable` is invoked. it must check the actual state
- we can just mention in javadoc that "spurious notifications" are possible 

BTW if we want to be future-proof and not lock ourselves into current (only) use-case, it would be more desirable to specify whether the `table` object is from before the change, or after. Or maybe provide both.
--  i am not advocating for this though. Let's do as little, and as simply as we can.

Reply:
Thank you for stressing on this.
I over-engineered the first version of this interface. I've refactored this interface to `TableInvalidationCallback` which actually is much easier to grasp and more specific to the purpose at hand for this PR.